yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,409 @@
+
+### Recombining for teardown
+
+The public kernel **tail** will recombine the `gas_used` values from the non-revertible and revertible phases.
+
+```
+kernel_public_inputs.gas_limits := public_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used =
+    public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+    + public_kernel_public_inputs.end.gas_used.da_gas_used
+
+kernel_public_inputs.end.gas_used.compute_gas_used :=
+    public_kernel_public_inputs.end_non_revertible.gas_used.compute_gas_used +
+    + public_kernel_public_inputs.end.gas_used.compute_gas_used
+
+
+fee_per_da_gas := public_kernel_public_inputs.constants.historical_header.global_variables.fee_per_da_gas
+fee_per_compute_gas := public_kernel_public_inputs.constants.historical_header.global_variables.fee_per_compute_gas
+
+kernel_public_inputs.transaction_fee :=
+    fee_per_da_gas * kernel_public_inputs.end.gas_used.da_gas_used
+    + fee_per_compute_gas * kernel_public_inputs.end.gas_used.compute_gas_used
+    + kernel_public_inputs.gas_limits.inclusion_fee
+```
+
+# Teardown
+
+After executing the teardown function, the teardown kernel will assert that the `transaction_fee` that was made available in the public context is equal to `kernel_public_inputs.transaction_fee`.
+
+This ensures that the `pay_fee` function on the enshrined gas token contract **can** be called with the correct amount.
+
+### Validating fee payment
+
+During execution of the teardown function, the public VM will need to inform the kernel **if** the `pay_fee` function was called.
+
+This is done via a `feePaid` boolean in the MachineState.
+
+This field is included in the `PublicCircuitPublicInputs` struct as `fee_paid`.
+
+The public kernel teardown will then set:
+
+```
+kernel_public_inputs.transaction_fee :=
+    public_inputs.fee_paid ? kernel_public_inputs.transaction_fee : 0
+```
+
+:::warning
+How does the VM know that the `pay_fee` function has been called?
+How do we convince the kernel that the machine state provided in the public inputs is correct?
+:::
+
+### Sequencer payouts
+
+Sequencers are paid in `$AZT` **on L1**.
+
+After a block has been published/verified, the sequencer is paid the sum of the `inclusion_fee`s in the block for all transactions with a non-zero `transaction_fee`.
+
+# Published data
+
+The [published transaction data](../data-publication-and-availability/published-data.md) includes:
+- `da_gas_used`
+- `da_gas_limit`
+- `max_fee_per_da_gas`
+- `compute_gas_used`
+- `compute_gas_limit`
+- `max_fee_per_compute_gas`
+- `inclusion_fee`
+- `transaction_fee`
+
+`fee_per_da_gas` and `fee_per_compute_gas` are not included in the published data per transaction because they are included in the block header.
+
+`transaction_fee` reflects the amount of `$AZT` that **was paid** for the transaction.
+
+# Kernel state diagram
+
+The current state diagram is as follows:
```

Comment:
I hate you for making this beautiful diagram and placing it only at the very end of the doc, when I could've used it a lot earlier :-P